Adding documentation on otel with vector (#835)
stackabletech/documentation: modules/concepts/pages/observability/telemetry.adoc

<1> Boolean: `true`, `false`
<2> String: `error`, `warn`, `info`, `debug`, `trace`, `off` (or {env-filter-syntax}[more complex filters,window=_blank])
<3> String: E.g. `https://my-collector:4317` (Note: it must contain the scheme)

== OpenTelemetry with Vector

> OpenTelemetry signals can be configured to behave like xref:logging.adoc[product logging]. Using the {vector-helm-chart}[vector helmChart] requires to configure `containerPorts` and `service.ports` as well as `customConfig.sources.otel`:

Reviewer: ```suggestion
OpenTelemetry signals can be configured to behave like xref:logging.adoc[product logging]. Using the {vector-helm-chart}[Vector Helm Chart] requires `containerPorts`, `service.ports`, and `customConfig.sources.otel` to be configured:
```

> OpenTelemetry signals can be configured to behave like xref:logging.adoc[product logging]. Using the {vector-helm-chart}[vector helmChart] requires to configure `containerPorts` and `service.ports` as well as `customConfig.sources.otel`:

Reviewer 2: What does "behave like product logging" mean exactly? A short explanation what that refers to, would be nice.

> OpenTelemetry signals can be configured to behave like xref:logging.adoc[product logging]. Using the {vector-helm-chart}[vector helmChart] requires to configure `containerPorts` and `service.ports` as well as `customConfig.sources.otel`:

Reviewer 2: ```suggestion
OpenTelemetry signals can be configured to behave like xref:logging.adoc[product logging]. Using the {vector-helm-chart}[Vector Helm Chart{external-link-icon}^] requires to configure the Helm values `containerPorts`, `service.ports`, and `customConfig.sources.otel`:
```

Reviewer 2: Adjusted the suggestion to include the external link indicator for the external Helm Chart

Author: Donen [0241521](https://github.com/stackabletech/documentation/pull/835/commits/02415215f616703f59f61a62d6bbcabd1d644042)

[source,yaml]
----
containerPorts:
  - {name: vector,    containerPort: 6000, protocol: TCP}
  - {name: otel-grpc, containerPort: 4317, protocol: TCP}
  - {name: otel-http, containerPort: 4318, protocol: TCP}
service:
  ports:
  - {name: vector,    port: 6000, targetPort: 6000, protocol: TCP}
  - {name: otel-grpc, port: 4317, targetPort: 4317, protocol: TCP}
  - {name: otel-http, port: 4318, targetPort: 4318, protocol: TCP}
customConfig:
  sources:
    otel:
      type: opentelemetry
      grpc:
        address: 0.0.0.0:4317
      http:
        address: 0.0.0.0:4318
----

> The `endpoint` in the operator's Helm values must point to the ports defined above.

Reviewer 2: ```suggestion
The `endpoint` Helm values for the telemetry exporters in the Stackable Operator Helm Chart must point to the otel-grpc ports defined above.
```

[source,yaml]
----
telemetry:
  otelLogExporter:
    enabled: true

>     endpoint: http://vector-aggregator.<namespace>.svc.cluster.local:4317

Reviewer 2: Actually, if there is a specific reason why this port was used and not otel-http for example, even though they are defined in Vector, feel free to elaborate why. If any of them can be used, maybe the above sentence can be adjusted to reflect that.

Author: The way I understood it, you can choose whatever port you want as long as you expose it in the aggregator.yaml shown above.

I understood that the gRPC standard port is 4317 and thus I kept the convention. Technically you could choose anything. I'm now not sure, I'd like to preserve some kind of convention and story throughout the guide. I think we could introduce `<gRPC-port>` and `<http-port>` if you like it better?

Reviewer 3: I think we should use the standard ports 4317 for gRPC, 4318 for HTTP. But yes they should be mentioned so the reader doesn't assume they are random.

Author: Done [0241521](https://github.com/stackabletech/documentation/pull/835/commits/02415215f616703f59f61a62d6bbcabd1d644042)

  otelTraceExporter:
    enabled: true
    endpoint: http://vector-aggregator.<namespace>.svc.cluster.local:4317
----

> === Normalizing operator logs to the Stackable log schema

Reviewer 2: Just a bit confused, above the sections starts with OpenTelemetry signals (but behaving like logs?), but here the subsection is about logs, so something different? So either that's something general for logs, so maybe needs to be a separate 2nd level section (==) or at least for me it needs a bit more of an explanation connecting the two concepts here.

Author: Good catch, hope it's better now: Adressed with [3c4d60f](https://github.com/stackabletech/documentation/pull/835/commits/3c4d60f62eacd6f28a80ae47b4b7dfa7e6e90f6a)

> Operators do not run vector sidecars. Using OpenSearch, a VRL transform that maps the OTLP fields onto the

Reviewer 2: ```suggestion
Stackable Operators do not run Vector sidecars. Using OpenSearch, a VRL transform that maps the OTLP fields onto the
```

> Operators do not run vector sidecars. Using OpenSearch, a VRL transform that maps the OTLP fields onto the

Reviewer 2: VRL? I think the abbrevation written out (or written out in parentheses next to it) would be more useful for people not knowing what that is (me :D)

Author: Vector Remap Language. I shall do, thanks

Author: Adressed with [3c4d60f](https://github.com/stackabletech/documentation/pull/835/commits/3c4d60f62eacd6f28a80ae47b4b7dfa7e6e90f6a)

xref:tutorials:logging-vector-aggregator.adoc#_watch_the_logs[Stackable log schema] is needed to avoid re-indexing:

> [source,yaml]

Reviewer 2: Are these Helm values for Opensearch now? (Because of "using Opensearch" above) Maybe this also needs a linked Helm Chart then or mentioning that it's our own Opensearch Operator Helm Chart? And a clarification where to set the values.

Author: Do you mean the following `customConfig` ?

If so, no. It's a map that defines how to map otel log fields so they fit the same schema as stackable products would look like.  Open search is just a very good example for a indexed backend. Since it's in our stack you are right that we might wanna hint it here. I'll fix it.

Author: Adressed with [3c4d60f](https://github.com/stackabletech/documentation/pull/835/commits/3c4d60f62eacd6f28a80ae47b4b7dfa7e6e90f6a)

----
customConfig:
  transforms:
    normalize_otel_logs:
      type: remap
      inputs:
        - otel.logs # <1>
      source: |
        service_name = get(.resources, ["service.name"]) ?? null
        .pod         = service_name # <2>
        .container   = service_name # <2>
        .namespace   = get(.resources, ["k8s.namespace.name"]) ?? null
        .logger      = string(.scope.name) ?? null # <3>
        .level       = string(.severity_text) ?? null
        .built_info  = object(.attributes) ?? {} # <4>
        .cluster     = null
        .role        = null
        .roleGroup   = null
        .file        = null
        .source_type = "opentelemetry"
        del(.severity_text)
        del(.severity_number)
        del(.resources)
        del(.attributes)
        del(.scope)
        del(.trace_id)
        del(.span_id)
        del(.flags)
        del(.observed_timestamp)
        del(.dropped_attributes_count)
  sinks:
    opensearch:
      inputs:
        - vector
        - normalize_otel_logs # <5>
----

<1> The `opentelemetry` source fans out into `otel.logs`, `otel.metrics`, and

>     `otel.traces`. Unconsumed sub-streams produce a startup warning.

Reviewer 2: ```suggestion
<1> The `opentelemetry` source type diverges into `otel.logs`, `otel.metrics`, and
    `otel.traces` signals. Unconsumed sub-streams produce a startup warning.
```

(Don't know if "signals" is the right word, but some kind of naming would sound better I think. Oh, or is sub-stream the correct term? Then feel free to use that instead of signal in the suggestion)

> <2> Operators emit `service.name` (e.g. `stackable-airflow-operator`) as their

Reviewer 2: Is `service.name` correct here? Or should it be `service_name` like in the line it references?

Author: It's `service.name` from the operator pov and only caught here as service_name as I don't want an attribute to the type service but rather something storing away that information

Reviewer 3: Many injestors can't handle the dotted names. Some of them automatically remap them with `_`, others don't (I think OpenSearch Dataprepper didn't).

It's a mess out there, until everyone starts properly accepting Semantic Conventions :D

    primary identity. It is mapped to both `.pod` and `.container` but can be
    adapted to individual needs.
<3> The OTLP `scope.name` contains the Rust module path (e.g.
    `kube_runtime::controller`) and maps to the `.logger` field.
<4> Log-record attributes (e.g. `built_info.*`) are collected into a single object
    to keep the top-level schema flat.
<5> Reference transformer output as sinks input.

[NOTE]
====

> Namespace information is currently not emitted by the operator

Reviewer: ```suggestion
Namespace information is currently not emitted by the operator.
```
